File: 141-LinkedListCycle/141-LinkedListCycle.py
# Last updated: 8/16/2026, 9:52:04 PM
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        # if not head or not head.next: return False

        # visited = []
        # pos = 0

        # curr = head
        # while curr.next:
        #     if curr not in visited:
        #         visited.append(curr)
        #         curr = curr.next
        #         pos += 1
        #     else:
        #         return True
        
        # return False

        # two pointers, hare and tortoise
        # if there is a cycle, slow will catch fast eventually
        slow, fast = head, head
        while slow and fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast: return True

        return False
        

File: 141-LinkedListCycle/141-LinkedListCycle.java
// Last updated: 8/16/2026, 9:52:08 PM
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    // public boolean hasCycle(ListNode head) {
    //     if(head == null) return false;
    //     Set<ListNode> visited = new HashSet<>();
    //     ListNode curr = head;
        
    //     while(curr != null) {
    //         if (visited.contains(curr)) {
    //             return true;
    //         } else {
    //             visited.add(curr);
    //         }
    //         curr = curr.next;
    //     }
    //     return false;
    // }

    // method 2 - two pointers, hare and tortoise
    // fast pointer moves 2 steps at a time. 
    // slow               1
    // if there is a cycle, slow will catch up the fast eventually
    public boolean hasCycle(ListNode head) {
        ListNode fast = head, slow = head;
        while(fast != null && fast.next != null && slow != null) {
            fast = fast.next.next;
            slow = slow.next;
            if(fast == slow) return true;
        }
        return false;
    }
}